astropy/units/structured.py

```diff
@@ -515,3 +517,40 @@ def __ne__(self, other):
             other = other.item()
 
         return self.item() != other
+
+
+def _structured_unit_like_dtype(unit: UnitBase | StructuredUnit, dtype: np.dtype) -> StructuredUnit:
+    """Make a `StructuredUnit` of one unit, with the structure of a `numpy.dtype`.
+
+    Parameters
+    ----------
+    unit : UnitBase
+        The unit that will be filled into the structure.
+    dtype : `numpy.dtype`
+        The structure for the StructuredUnit.
+
+    Returns
+    -------
+    StructuredUnit
+    """
+    # Make sure dtype is structured
```

Comment:
I think for a private function we should be able to assume that the input is correct (i.e., just let the `AttributeError` or whatever else happen if not). python is not LBYL.

Reply:
I stuck this in for type narrowing since type hints are LBYL. There's no way to indicate that a dtype is structured except by checking. Hopefully NumPy will parametrize dtype as a `Generic[...]` or have a ``StructuredDtype`` protocol.

We could add that protocol...

```python
@runtime_checkable
class StructuredDtype(Protocol):
    @property
    def names(self) -> tuple: ...
    @property
    def fields(self) -> Mapping: ...
```

Reply:
Understood, but can you nevertheless remove it? It is one thing to label things to help people and computers understand intent, quite another to make the code slower and more complex... I find it a beauty of python that one does not presume to know beforehand all possible input one might get.

Reply:
Probably more ``typing`` than Astropy wants right now, but the suggested Protocol should solve both problems. This function knows that the dtype is structured and there's no checks necessary.

Reply:
Yes, having that `Protocol` would indeed solve it.  And could live in separate typing files, where it wouldn't bug me :smiley_cat:  More seriously, something to suggest to numpy?

Reply:
https://github.com/numpy/numpy/issues/22286

In the meantime... I can drop in a ``_typing.py`` file where we can start playing around with typing in ``units``.

Reply:
✅

Reply:
Darn. https://docs.python.org/3/library/typing.html#typing.runtime_checkable

Reply:
Forgot it does not do ANY signature checking. The protocol will not work.

Reply:
That's a pity.  But let's leave all of this out of this PR. As is, this private function is only called with well-defined input so no checks should be necessary.

Reply:
https://github.com/numpy/numpy/issues/22286